### PR TITLE
[dynamo] FSDP + AC + torch.compile

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103953

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

### CODE REVIEW

File: test/distributed/test_dynamo_distributed.py [hunk 1]
```diff
@@ -291,13 +319,49 @@ def test_fsdp_inductor(self):
             outputs = fsdp_m(inputs)
             self.assertTrue(same(correct_outputs, outputs))
 
+
+    @skip_if_lt_x_gpu(1)
+    @unittest.skipIf(not has_triton(), "Inductor+gpu needs triton and recent GPU arch")
+    def test_fsdp_activation_checkpointing(self):
+        from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import (
```
Comment (wconstab) [inline test/distributed/test_dynamo_distributed.py:326]: do we want to also test the native torch.utils one too?

File: test/distributed/test_dynamo_distributed.py [hunk 2]
```diff
@@ -291,13 +319,49 @@ def test_fsdp_inductor(self):
...
+
+        with _dynamo_dist_per_rank_init(self.rank, self.world_size):
+            model, inputs = get_toy_model_for_activation_checkpointing(f"cuda:{self.rank}")
+            is_inner = lambda module: isinstance(module, ToyInnerModel)  # noqa: E731
+            wrap_policy = functools.partial(lambda_auto_wrap_policy, lambda_fn=is_inner)
+            model = apply_fsdp(model, wrap_policy, is_inner)
+            correct_outputs = model(inputs)
+            opt_model = torch.compile(model)
+            outputs = opt_model(inputs)
+            self.assertTrue(same(correct_outputs, outputs))
```
Comment (wconstab) [inline test/distributed/test_dynamo_distributed.py:356]: should we assert for an expected number of graphs (1?) to confirm no graph-breaks happened?

File: test/distributed/test_dynamo_distributed.py [hunk 3]
```diff
@@ -291,13 +319,49 @@ def test_fsdp_inductor(self):
...
+                )
+                apply_activation_checkpointing(
+                    model, checkpoint_wrapper_fn=checkpoint_wrapper_fn, check_fn=checkpoint_policy,
+                )
+            return model
+
+        with _dynamo_dist_per_rank_init(self.rank, self.world_size):
+            model, inputs = get_toy_model_for_activation_checkpointing(f"cuda:{self.rank}")
+            is_inner = lambda module: isinstance(module, ToyInnerModel)  # noqa: E731
+            wrap_policy = functools.partial(lambda_auto_wrap_policy, lambda_fn=is_inner)
```
Comment (wconstab) [inline test/distributed/test_dynamo_distributed.py:351]: are there any valid cases where we don't use the same policy for fsdp/ac that we need to test? cc @awgu

File: test/distributed/test_dynamo_distributed.py (hunk 2 above)
Comment (anijain2305) [reply test/distributed/test_dynamo_distributed.py:356]: Good call. Let me see if there is a way to get that info out and check it.

File: test/distributed/test_dynamo_distributed.py (hunk 1 above)
Comment (anijain2305) [reply test/distributed/test_dynamo_distributed.py:326]: Thats already tested in test/dynamo/test_activation_checkpointing.py

They are also tested heavily on HF models - more data here - https://github.com/pytorch/pytorch/pull/102935

File: torch/_dynamo/variables/torch.py [hunk 4]
```diff
@@ -895,7 +900,11 @@ def add_subgraph(name, gm):
                     next_name = candidate
 
             gm.__name__ = next_name
-            src = NNModuleSource(GetItemSource(self.source, next_name))
+            if self.source.guard_source().is_fsdp_module():
```
Comment (wconstab) [inline torch/_dynamo/variables/torch.py:903]: what does 'add_subgraph' do generally? are there more than one subgraph per higher order operator?

File: test/distributed/test_dynamo_distributed.py (hunk 2 above)
Comment (anijain2305) [reply test/distributed/test_dynamo_distributed.py:365]: Added a test that 
* Checks that there are 2 graphs - one for each FSDP wrapped module
* Check that the graph itself has a subgraph of activation checkpointing

File: torch/_dynamo/variables/torch.py (hunk 4 above)
Comment (anijain2305) [reply torch/_dynamo/variables/torch.py:903]: It is just a util that inserts the subgraph into the output graph module of Dynamo. Yes, there is no limit on how many subgraphs we can have. torch.cond, for example, has 2 - one for true branch and one for false.

In the case of activation checkpointing, there is only 1.

File: torch/_dynamo/utils.py [hunk 5]
```diff
@@ -1678,6 +1678,12 @@ def defake(x):
 # NB: The dictionary has to be created lazily after TorchPatcher is called so
 # that we pick up the disabled torch.utils.checkpoint wrapper. Therefore, it is
 # sitting in a separate function.
+# We also need the original untouched/ not disabled torch utils checkpoint
+# becuase distributed checkpointed wrappers import these utils before
+# TorchDynamo TorchPatcher runs.
+untouched_torch_utils_checkpoint = torch.utils.checkpoint.checkpoint
```
Comment (wconstab) [inline torch/_dynamo/utils.py:1684]: i guess this is one more straw on the back of 'lets overhaul skipfiles/disable? seems pretty hard to maintain and reason about this logic.  i'm not sure if there is a better way to go forward now.

File: torch/_higher_order_ops/wrap.py [hunk 6]
```diff
@@ -47,7 +47,7 @@ class TagActivationCheckpoint(HigherOrderOperator):
     """
 
     def __init__(self):
-        super().__init__("wrap_activation_checkpoint")
+        super().__init__("tag_activation_checkpoint")
```
Comment (wconstab) [inline torch/_higher_order_ops/wrap.py:50]: what is this? and why is 'wrap_activation_checkpoint' still left in the case statement in torchvariable?

Comment (wconstab) [reply torch/_higher_order_ops/wrap.py:50]: oh i see WrapActivationCheckpoint still exists.  Is WrapActivationCheckpoint doing things differently, not by passing information to aot about what to partition?

Comment (anijain2305) [reply torch/_higher_order_ops/wrap.py:50]: This is just something I missed in the last PR. We have both wrap and tag activation_checkpoint higher order ops right now. This is just a temporary stage. 

wrap_activation_checkpoint - This asks AOT Autograd to trace torch.utils.checkpoint code.
tag_activation_checkpoint - This just tags the nodes to be recomputable, and then partitioner actually duplicates the nodes. We avoid tracing through the torch.utils.checkpoint. This is currently preferred because here we use inductor functionalization for rng states.

Comment (anijain2305) [reply torch/_higher_order_ops/wrap.py:50]: This points to the other HigherOrderOp - https://github.com/pytorch/pytorch/pull/103953/files#diff-bc653d54b1dbf34241875a12617c6a2a86be63bce7543cca65bd38d7ed2d5afeL24

File: torch/_dynamo/utils.py (hunk 5 above)
Comment (anijain2305) [reply torch/_dynamo/utils.py:1684]: I know this is getting out of hand. I could not come up with a cleaner way for this right now.

File: torch/_functorch/partitioners.py [hunk 7]
```diff
@@ -473,18 +473,7 @@ def insert_node_in_graph(node):
...
-    for output in output_node.args[0]:
-        if isinstance(output, torch.fx.node.Node):
-            if output not in env:
-                print("output gradient dependent only on fwd", output)
-            new_outputs.append(insert_node_in_graph(output))
-        else:
-            new_outputs.append(output)
-
-    new_graph.output(new_outputs)
+    # The output node is already built by the traversal.
```
Comment (wconstab) [inline torch/_functorch/partitioners.py:476]: hmm, so is this a required change for correctness, or is it more of a BE cleanup (the traversal handles this but we were doing it up front instead)

**Review by wconstab — COMMENTED** [review]
i think it looks good. Would like to get @wanchaol to also review

File: torch/_functorch/partitioners.py (hunk 7 above)
Comment (anijain2305) [reply torch/_functorch/partitioners.py:476]: Its a BE cleanup.

File: torch/_higher_order_ops/wrap.py (hunk 6 above)
Comment (wconstab) [reply torch/_higher_order_ops/wrap.py:50]: got it. please also add that explanation of the 2 behaviors as a code comment, sounds useful.  (you almost have it, you have the half about what tagAC does but not as clear about the wrapAC one

File: test/distributed/test_dynamo_distributed.py [hunk 8]
```diff
@@ -313,14 +390,69 @@ def apply_fsdp(model, wrap_policy):
...
+    @patch.object(torch._inductor.config, "fallback_random", True)
+    def test_hf_bert_fsdp_activation_checkpointing(self):
+        from transformers.models.bert.modeling_bert import BertLayer
+        from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import (
+            apply_activation_checkpointing,
+            checkpoint_wrapper,
+            CheckpointImpl,
+        )
+
+        def apply_fsdp(model, wrap_policy, use_activation_checkpointing=True):
```
Comment (wanchaol) [inline test/distributed/test_dynamo_distributed.py:429]: would it make sense to make this a common util, as we defined the same thing in two different places?

File: torch/_dynamo/variables/torch.py [hunk 9]
```diff
@@ -1222,7 +1230,11 @@ def speculate_branch(branch):
             example_value = deepcopy_to_fake_tensor(example_res, tx.fake_mode)
 
             p_args = (lowered_node,) + p_args
-        elif self.value.__name__ in ("wrap", "wrap_activation_checkpoint"):
+        elif self.value.__name__ in (
+            "wrap",
+            "wrap_activation_checkpoint",
+            "tag_activation_checkpoint",
```
Comment (wanchaol) [inline torch/_dynamo/variables/torch.py:1236]: curious where do we generate these `wrap` or `tag` ac calls inside dynamo? is the dynamo detecting `torch.utils.checkpoint` and convert this call to a highorderop?

**Review by wanchaol — APPROVED** [review]
lgtm, the CI failure looks correlated https://hud.pytorch.org/pr/pytorch/pytorch/103953#14451950178

File: test/distributed/test_dynamo_distributed.py [hunk 10]
```diff
@@ -313,14 +390,69 @@ def apply_fsdp(model, wrap_policy):
...
+            ):
+                print(f"Running hf_bert test for {test_instance}")
+                model, inputs = get_hf_bert(self.rank)
+                reset_rng_state()
+                eager_model = apply_fsdp(model, wrap_policy)
+                correct_outputs = eager_model(**inputs)
+                correct_loss = correct_outputs.loss
+                correct_loss.backward()
+
+                reset_rng_state()
```
[the remark this answers is not in the document]
Comment (anijain2305) [reply test/distributed/test_dynamo_distributed.py:402]: Yes, its only for accuracy testing.

File: torch/_dynamo/variables/torch.py (hunk 9 above)
Comment (anijain2305) [reply torch/_dynamo/variables/torch.py:1236]: Yes, we do this inside Dynamo. This PR is actually doing exactly that. It detects torch.utils.checkpoint and converts it into higherorderop - https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/utils.py#L1681

File: test/distributed/test_dynamo_distributed.py (hunk 8 above)
Comment (anijain2305) [reply test/distributed/test_dynamo_distributed.py]: Good call. Done!

File: torch/_higher_order_ops/wrap.py (hunk 6 above)
Comment (anijain2305) [reply torch/_higher_order_ops/wrap.py:66]: Done!

**Comment by voznesenskym** [comment]
Do you mind testing this against https://github.com/pytorch/pytorch/pull/103711 ? I know its an onerous ask, but I want to make sure we also keep in mind the relative future direction of PT2 + FSDP for things like this.

**Comment by anijain2305** [comment]
> Do you mind testing this against #103711 ? I know its an onerous ask, but I want to make sure we also keep in mind the relative future direction of PT2 + FSDP for things like this.

Oh missed this message. Will certainly do in next couple of weeks. Trying to wrap up some work, and will pick this up.
